### PR TITLE
Track events inside OBW and checklist

### LINKED ISSUE #85: Add Track Events
It appears that we can now use Jetpack to record non-`jetpack` prefixed events! Awesome! There are two helper utils for making recording tracks in our plugin quite easy:

- PHP tracks client [usage example](https://github.com/Automattic/jetpack/blob/f8078c2cd12ac508334da2fb08e37a92cf283c14/_inc/lib/tracks/client.php#L7-L14) enables using `jetpack_tracks_record_event()` from within PHP
- [JS Ajax client](https://github.com/Automattic/jetpack/blob/f8078c2cd12ac508334da2fb08e37a92cf283c14/_inc/lib/tracks/tracks-ajax.js#L6)  provides an ajax helper for recording tracks. I think using the approach we have setup in wc-admin might be a bit nicer from JS, but lets use what we have here via Jetpack.

I've created a spreadsheet for us to add in all the tracks we are recording as part of this project. I have yet to create entries for all the task list steps, but lets just update the spreadsheet as we go along:

[https://docs.google.com/spreadsheets/d/17yZeaWktUSq_lf6fVhe61hyTCXO_09gKmqXdcUzGj18/edit?usp=sharing](https://docs.google.com/spreadsheets/d/17yZeaWktUSq_lf6fVhe61hyTCXO_09gKmqXdcUzGj18/edit?usp=sharing)

I opted to standardize on the `atomic_` prefix for these track names, since it is already whitelisted, and these are indeed all atomic sites. /cc @jhnstn to make sure that is an okay namespace to use for tracking events in this project.

Note: To get Jetpack tracks helpers working in your local install, you might need to manually set the following site option:

`update_option( 'jetpack_tos_agreed', true );`

### PR DESCRIPTION
**Pull request by joshuatf**
Fixes #85 

This PR adds event tracking for the following events:
* OBW step completion
* Tasklist item clicked
* Tasklist finished (clicked "I'm done")
* Edit address in OBW clicked
* Calypsoify manually turned on/off

### Screenshots
screen shot 2018-11-01 at 3 41 10 pm

### Testing
1.  Make sure you have a Jetpack connected site with tracking enabled.
2.  Try the various above events to make sure they come through (may take a few minutes to show up) - https://mc.a8c.com/tracks/live/?eventname=%25atomic_%25&user=&useragent=

### Notes
* The javascript events are prefixed automatically with `jetpack_`
* The task list completion event is dependent on https://github.com/Automattic/wc-calypso-bridge/pull/130

### CODE REVIEW

File: class-wc-calypso-bridge.php [hunk 1]
```diff
@@ -91,9 +92,31 @@ public static function instance() {
...
+	 * Track Calypsoify events when turned on or off
+	 */
+	public function track_calypsoify_toggle() {
+		if ( isset( $_GET['calypsoify'] ) ) { // WPCS: CSRF ok.
+			$current_user      = wp_get_current_user();
+			$calypsoify_status = (int) get_user_meta( $current_user->ID, 'calypsoify', true );
+			if ( 1 === $calypsoify_status && 0 === (int) $_GET['calypsoify'] ) { // WPCS: CSRF ok.
+				jetpack_tracks_record_event(
+					$current_user,
+					'atomic_wc_calypsoify_off'
```
Comment (justinshreve) [inline class-wc-calypso-bridge.php]: While not explicitly addressed in the "Tracks Event Design & Naming Conventions" field guide page, there is a section in the FG that talks about not using variables in the name such as post type, and instead use a single event with a prop. I feel like that can apply here. 

So I'm wondering if maybe we should name this `atomic_wc_calypsoify_toggle` instead with a prop for off/on.

File: assets/js/calypsoify.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    /**
+     * Record checklist task click
+     */
+    $( '.checklist__task-title a, .checklist__task-secondary a' ).click( function() {
+        var $task = $( this ).closest( '.checklist__task' )
+        var status = $task.hasClass( 'is-completed' ) ? 'complete' : 'incomplete';
+        var taskTitle = $task.data('title');
+
+        window.jpTracksAJAX.record_ajax_event(
+            'atomic_wc_tasklist_clicked',
```
Comment (justinshreve) [inline assets/js/calypsoify.js]: This should be `atomic_wc_tasklist_click`

> use a singular noun or present tense verb

File: assets/js/calypsoify.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+     * Track 'I'm done' completion on task list
+     */
+    $( '.setup-footer a' ).click( function(e) {
+        e.preventDefault();
+        var progressNumber = $( '.checklist__header-progress-number' ).text().split( '/' );
+        var complete = progressNumber[0];
+        var total = progressNumber[1];
+        var percentage = parseFloat( complete / total ).toFixed( 2 ) * 100;
+        window.jpTracksAJAX.record_ajax_event(
+            'atomic_wc_tasklist_finished',
```
Comment (justinshreve) [inline assets/js/calypsoify.js]: This should be `atomic_wc_tasklist_finish`

> use a singular noun or present tense verb

File: assets/js/calypsoify.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
+( function( $ ) {
+    'use strict';
+
+    /**
+     * Record checklist task click
+     */
+    $( '.checklist__task-title a, .checklist__task-secondary a' ).click( function() {
+        var $task = $( this ).closest( '.checklist__task' )
+        var status = $task.hasClass( 'is-completed' ) ? 'complete' : 'incomplete';
+        var taskTitle = $task.data('title');
```
Comment (justinshreve) [inline assets/js/calypsoify.js]: We might want to add a normalized, untranslated id to the array of task fields instead and use that here (or both).

I know for the initial launch none of these tasks are translated, but in the future the task titles could be changed/translated and we'll still want to associate those events together.

File: assets/js/calypsoify.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        var complete = progressNumber[0];
+        var total = progressNumber[1];
+        var percentage = parseFloat( complete / total ).toFixed( 2 ) * 100;
+        window.jpTracksAJAX.record_ajax_event(
+            'atomic_wc_tasklist_finished',
+            'click',
+            { 
+                complete: complete,
+                total: total,
+                percentage: percentage
```
Comment (justinshreve) [inline assets/js/calypsoify.js]: Great idea passing the percentage here! 👍

File: class-wc-calypso-bridge.php [hunk 6]
```diff
@@ -91,9 +92,31 @@ public static function instance() {
...
+				);
+			} elseif ( 0 === $calypsoify_status && 1 === (int) $_GET['calypsoify'] ) { // WPCS: CSRF ok.
+				jetpack_tracks_record_event(
```
Comment (justinshreve) [inline class-wc-calypso-bridge.php]: It looks like we need to manually pass the `blog_id` prop for PHP events. You can get the blog id from `Jetpack_Options::get_option( 'id' );`.

The Javascript events already seem to include it.

File: class-wc-calypso-bridge.php (hunk 1 above)
Comment (joshuatf) [reply class-wc-calypso-bridge.php]: That seems like a good idea plus we can clean up the logic here a bit.  https://github.com/Automattic/wc-calypso-bridge/pull/133/commits/001ffba4bc7f194019218b4ee1db1322d09ab74b

File: assets/js/calypsoify.js (hunk 2 above)
Comment (joshuatf) [reply assets/js/calypsoify.js]: Updated!

File: assets/js/calypsoify.js (hunk 3 above)
Comment (joshuatf) [reply assets/js/calypsoify.js]: Updated!

File: assets/js/calypsoify.js (hunk 4 above)
Comment (joshuatf) [reply assets/js/calypsoify.js]: Ah good thinking!  Let me know if this looks reasonable - https://github.com/Automattic/wc-calypso-bridge/pull/133/commits/6e076c1a6b2ad3aea6cf13ba9c847b790a193f23

**Comment by timmyc** [comment]
This is looking great @joshuatf! The only thing I'm wondering about is if we should first do a null check on `window.jpTracksAJAX` to ensure it is defined. While _highly unlikely_ this will ever be undefined in the context of these AT sites, if somehow the site gets disconnected from Jetpack, I believe that particular global is not defined.

File: includes/class-wc-calypso-bridge-admin-setup-wizard.php [hunk 7]
```diff
@@ -107,6 +107,15 @@ public function setup_wizard() {
 
 		// @codingStandardsIgnoreStart
 		if ( ! empty( $_POST['save_step'] ) && isset( $this->steps[ $this->step ]['handler'] ) ) {
+			$current_user = wp_get_current_user();
+			jetpack_tracks_record_event(
```
Comment (timmyc) [inline includes/class-wc-calypso-bridge-admin-setup-wizard.php]: I suppose if we are being ultra paranoid around checking to see if the JS helper is defined we should check here too.

File: assets/js/calypsoify.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
+( function( $ ) {
+    'use strict';
+
+    /**
```
Comment (timmyc) [inline assets/js/calypsoify.js]: I suppose we could just wrap these in a `if (window.jpTracksAJAX)` but i'm likely being too paranoid :)

Comment (joshuatf) [reply assets/js/calypsoify.js]: Seems like a safe idea 😄

File: includes/class-wc-calypso-bridge-admin-setup-wizard.php (hunk 7 above)
Comment (joshuatf) [reply includes/class-wc-calypso-bridge-admin-setup-wizard.php]: Moved this into its own function to dry things up a bit - https://github.com/Automattic/wc-calypso-bridge/pull/133/commits/9e06bcf64eba90d75012341b77a9ffc69d92bc25 . Let me know what you think.

**Comment by justinshreve** [comment]
> Cleaned up a bit and made those changes. Let me know if you think we need to change anything else here.

Looking good @joshuatf! I left one more comment about the id naming here:
https://github.com/Automattic/wc-calypso-bridge/pull/133#discussion_r230413976

**Comment by joshuatf** [comment]
@justinshreve Just pushed a change for the IDs.  Let me know how we're looking now!

File: class-wc-calypso-bridge.php [hunk 9]
```diff
@@ -108,6 +110,42 @@ public function check_setup_param() {
...
+		}
+	}
+
+	/**
+	 * Record event using JetPack if enabled
+	 *
+	 * @param string $event_name Name of the event.
+	 * @param array  $event_params Custom event params to capture.
+	 */
+	public static function record_event( $event_name, $event_params ) {
```
Comment (justinshreve) [inline class-wc-calypso-bridge.php]: Nice 👍

**Comment by joshuatf** [comment]
Awesome!  Thanks for the all the input @justinshreve and @timmyc 😃
